internal/docstore/firedocstore/fs.go

```diff
@@ -12,26 +12,150 @@
+	"gocloud.dev/gcp"
 	"gocloud.dev/internal/docstore"
 	"gocloud.dev/internal/docstore/driver"
 	"gocloud.dev/internal/gcerr"
+	"gocloud.dev/internal/useragent"
+	"google.golang.org/api/option"
 	pb "google.golang.org/genproto/googleapis/firestore/v1"
 	"google.golang.org/grpc/metadata"
 )
 
+func init() {
+	docstore.DefaultURLMux().RegisterCollection(Scheme, &lazyCredsOpener{})
+}
+
+type lazyCredsOpener struct {
+	init   sync.Once
+	opener *URLOpener
+	err    error
+}
+
+func (o *lazyCredsOpener) OpenCollectionURL(ctx context.Context, u *url.URL) (*docstore.Collection, error) {
+	o.init.Do(func() {
+		creds, err := gcp.DefaultCredentials(ctx)
+		if err != nil {
+			o.err = err
+			return
+		}
+		client, _, err := Dial(ctx, creds.TokenSource)
+		if err != nil {
+			o.err = err
+			return
+		}
+		o.opener = &URLOpener{Client: client}
+	})
+	if o.err != nil {
+		return nil, fmt.Errorf("open collection %s: %v", u, o.err)
+	}
+	return o.opener.OpenCollectionURL(ctx, u)
+}
+
+// Dial returns a client to use with Firestore and a clean-up function to close
+// the client after used.
+func Dial(ctx context.Context, ts gcp.TokenSource) (*vkit.Client, func(), error) {
+	c, err := vkit.NewClient(ctx, option.WithTokenSource(ts), useragent.ClientOption("docstore"))
+	return c, func() { c.Close() }, err
+}
+
+// Scheme is the URL scheme firedocstore registers its URLOpener under on
+// docstore.DefaultMux.
+const Scheme = "firedocstore"
+
+// URLOpener opens firedocstore URLs like
+// "firedocstore://myproject/mycollection?name_field=myID".
+//
+//   - The URL's host holds the GCP projectID.
+//   - The only element of the URL's path holds the path to a Firestore collection.
+// See https://firebase.google.com/docs/firestore/data-model for more details.
+//
+// The following query parameters are supported:
+//
+//   - name_field (required): firedocstore requires that a single string field,
+// name_field, be designated the primary key. Its values must be unique over all
+// documents in the collection, and the primary key must be provided to retrieve
+// a document.
+type URLOpener struct {
+	// Client must be set to a non-nil client authenticated with Cloud Firestore
+	// scope or equivalent.
+	Client *vkit.Client
+}
+
+// OpenCollectionURL opens a docstore.Collection based on u.
+func (o *URLOpener) OpenCollectionURL(ctx context.Context, u *url.URL) (*docstore.Collection, error) {
+	q := u.Query()
+
+	nameField := q.Get("name_field")
+	if nameField == "" {
+		return nil, errors.New("open collection %s: name_field is required to open a collection")
+	}
+	q.Del("name_field")
+	for param := range q {
+		return nil, fmt.Errorf("open collection %s: invalid query parameter %q", u, param)
+	}
+	cn, err := collectionNameFromURL(u)
+	if err != nil {
+		return nil, fmt.Errorf("open collection %s: %v", u, err)
+	}
+	return OpenCollection(o.Client, cn.ProjectID, cn.CollPath, nameField), nil
+}
+
+// CollectionName identifies a collection name.
```

Comment:
I would leave this out. It's exported, but not usable anywhere. Just parse the individual fields in the `URLOpener` as we do elsewhere.

#1860 discusses making compound ID handling more consistent across the board.

Reply:
Done.